Urdf with args and yaml configuration

dhled opened the pull request:
Move the duplicates parameters into common file and uses yaml file and xacro macro to configure urdf.

This PR includes:

- Refactoring the common ur macro in a common files
- Parse the parameters from yaml files
- Put every ur specific urdf into a common one
- Modify launch files and bringups according to the new parameters

Modifications:
- 3 new yaml files to configure the robots parameters one for each robot model / urdf (joints_limits, physical, visual)
- ur.common.xacro contains the parsing of the yaml files
- ur.geometry contains the geometry of the robot
- DOESN'T CONTAINS JOINT LIMITED ANYMORE
- Bringups launch file and tests changes accordingly.

This PR doesn't apply to ur-e series (WIP in another PR)

gavanderhoorn commented:
Nice to see some work in this direction.

Some first suggestions / comments:

 - give the `.yaml` files a default location, and use default args in the xacro macro to load the files from there. Only if users want to use custom settings they'd have to provide an alternative location for the `.yaml` file. That would remove the need for the `urN_robot_yaml.urdf.xacro` files completely, which would add yet another (confusing) set of xacros to the package.

 - (this is a suggestion / debatable): do away with the in-xacro constants completely: so all D-H derived constants, all limits, mesh locations, link lengths, everything. Place all of them in `.yaml` files, one for each robot variant. Rationale: preparation for when we can use the calibration data from each individual UR in the xacro (#357) and we essentially end up with a single macro for all robots that gets parameterised by the data in the `.yaml` files. This could allow us to simplify things significantly.

 - if we do this: also remove the limit args on the `urN_robot` macros: all limits can be set using the `.yaml` files, no need for the args

 - if we don't migrate all constants to the `.yaml` file, let's not make the mistake of including the `world` frame in the main xacro files agian (#284).

 - iirc, we don't need to limit *all* joints to `[-pi, pi]`, just one (the `elbow_joint`). See #268.

gavanderhoorn commented:
@ipa-led: as I used a squash-and-merge of #356 the first few commits of your PR now conflict with `kinetic-devel`. Sorry about that.

Could you rebase and the fix merge conflicts? That would also give you an opportunity to drop all the merge commits that Github introduced when updating your branch.

dhled commented:
Hey @gavanderhoorn 

Sorry for the delay on this one. I've made the requested changes. Please tell what you think.

If possible @ipa-nhg or @ipa-jfh can review it too.

gavanderhoorn commented:
Starting to get there @ipa-led :)

Any reason you did not add the mesh URLs to the yaml files?

That would allow ppl to use custom meshes, which would help with #267.

gavanderhoorn commented:
Similar question on the inertia parameters and `origin`s.

dhled commented:
No particular reason. It looks, indeed, that we can add more parameters (we also discussed this with Nadia). I will try to update it next week. Thank you.

v4hn commented:
This looks like a nice change w.r.t. urdf-calibration support.

I'm not sure whether it makes sense to export *everything* to yaml, because you will eventually end up with a general-purpose xacro forward to yaml... unless this is exactly what you want.
Might make it harder to debug things though when setting up new systems.

But I do see the advantage to overload the specific mesh of wrist3...

nitpick: typo `compute_lenght`

gavanderhoorn commented:
@v4hn wrote:
> I'm not sure whether it makes sense to export everything to yaml, because you will eventually end up with a general-purpose xacro forward to yaml... unless this is exactly what you want.

yes, that is one of the possible end goals.

But nothing will be removed without leaving some time for testing.

ipa-nhg commented on ur_description/urdf/ur10.urdf.xacro:
```diff
@@ -92,12 +40,8 @@
-        <limit lower="${shoulder_pan_lower_limit}" upper="${shoulder_pan_upper_limit}" effort="330.0" velocity="2.16"/>
-      </xacro:if>
+      <limit lower="${shoulder_pan_lower_limit}" upper="${shoulder_pan_upper_limit}"
+             effort="${shoulder_pan_effort_limit}" velocity="shoulder_pan_velocity_limit"/>
```
for the velocity limit the curly brackets are missed:

> velocity="**${** shoulder_pan_velocity_limit **}**"/>

ipa-nhg commented on ur_description/config/ur5_default.yaml:
```diff
@@ -0,0 +1,88 @@
+
+offsets:
+  shoulder_offset: 0.13585 # measured from model
+  elbow_offset: -0.119 # measured from model
```
the elbow_offset was originally -0.119**7**

ipa-nhg commented on ur_description/config/ur5_joint_limited.yaml:
```diff
@@ -0,0 +1,88 @@
+
+offsets:
+  shoulder_offset: 0.13585 # measured from model
+  elbow_offset: -0.119 # measured from model
```
the elbow_offset was originally -0.119**7**

ipa-nhg commented:
Great job @ipa-led !! 

_From the technical point of view:_

I tested the simulation of the 3 robots and it is working if you solve the [issue](https://github.com/ros-industrial/universal_robot/pull/371#discussion_r219189507)  I commented.

Also I generated the urdf code of the 3 models using:
`rosrun xacro xacro urX_robot.urdf.xacro > urX_robot.urdf`

and I compared the outcome with the outcome of the current upstream version, the only differences I found were a mistake copying the [numbers](https://github.com/ros-industrial/universal_robot/pull/371#discussion_r219190263), that the world link was removed and that all the `xacro:property` are not listed.

_About the style_:

I really like this solution, it is much cleaner and give us the chance to simplify the repository.

dhled commented:
Made the change requested by @ipa-nhg 

I will export also the rest of the parameters.

gavanderhoorn commented:
This is not finished iiuc @ipa-nhg.

ipa-nhg commented:
@gavanderhoorn what about merge this PR and create a new PR for the new changes?
The problem I see is a future conflict with https://github.com/ros-industrial/universal_robot/pull/380/files

gavanderhoorn commented:
> @gavanderhoorn what about merge this PR and create a new PR for the new changes?
> The problem I see is a future conflict with https://github.com/ros-industrial/universal_robot/pull/380/files

#380 introduces new packages, so none of the files that this PR touches are changed by #380.

ipa-nhg commented:
@gavanderhoorn I'm agree that #380 introduce new packages, but those packages have the same function as the modified here and have to be updated to follow the same structure

gavanderhoorn commented:
Is this still something you'd like to continue with @ipa-led?

dhled commented:
On it for the wrid19

dhled commented:
Should sync with #414
Not passing tests because it break the CI tests.

fmauch commented:
Wouldn't it make sense to also put the kinematics stuff into `ur_common.xacro`? They seem to be the same for all robots.

fmauch commented on ur_description/urdf/ur.common.xacro:
```diff
@@ -0,0 +1,160 @@
+    <xacro:property name="wrist_3_colision_mesh" value="${__wrist_3_mesh['collision']['mesh']}" scope="parent"/>
+  </xacro:macro>
+
+  <xacro:macro name="compute_lenght" params="d1 a2 a3 d4 d5 d6 shoulder_offset elbow_offset">
```
There's a small typo. It should be `compute_length` I guess.

fmauch commented:
@ipa-led did you leave out the e-series on purpose?

fmauch commented on ur_description/urdf/ur.common.xacro:
```diff
@@ -0,0 +1,160 @@
+    <xacro:property name="base_visual_mesh" value="${__base_mesh['visual']['mesh']}" scope="parent"/>
+    <xacro:property name="base_visual_material_name" value="${__base_mesh['visual']['material']['name']}" scope="parent"/>
+    <xacro:property name="base_visual_material_color" value="${__base_mesh['visual']['material']['color']}" scope="parent"/>
+    <xacro:property name="base_colision_mesh" value="${__base_mesh['collision']['mesh']}" scope="parent"/>
```
The `colision` typo is going through all the places :)

Commit:
deleted ur_e specific repositories
* moved ur_e_description meshes files to ur_description

jrgnicho reviewed (changes requested):
Resolve the conflicts

fmauch commented on ur_description/urdf/ur3_robot.urdf.xacro:
```diff
@@ -2,25 +2,18 @@
-    transmission_hw_interface="$(arg transmission_hw_interface)"
-  />
-
-  <link name="world" />
```
I think, we shouldn't remove this in the standalone file. Isn't it quite common practice to have a x_robot.urdf.xacro with a world link in there?

At least for gazebo this seems to make a huge difference, as only then the parameter to lift the robot into the air seems to be evaluated. With the current implementation the gazebo models start shaking immediately.

gavanderhoorn replied:
the issue here is that the file used to mix "being" a regular top-level xacro file with adding Gazebo compatibility.

As discussed this afternoon, it'd be best if the Gazebo-specific parts here are all moved to a separate set of files, and then using `gazebo` link references to add additional properties where needed.

Having a `world` link in this file can actually be quite problematic. See also #284.

fmauch replied:
thanks for pointing that out :)

dhled replied:
I agree with @gavanderhoorn but the PR starts to be a bit sized, should we move to a dedicated  branch "reshape urdf" to handle everything on the reshape of the urdf ?

fmauch replied:
Then we should update the gazebo launch files accordingly. With the current situation the robot is standing on the ground which will cause it to have collisions.

In the launchfile for ur_gazebo we add a parameter `-z 0.1` to lift it from the ground, but that only seems to work if a world link is present (correct me, if I'm wrong here).

So we could introduce a file `ur3_robot_gazebo.urfd.xacro` with the world link present?

dhled commented:
New update:
Reshapped a bit the commits.list to be more understandable
Fixed the inertia to match the one of @fmauch in #426

dhled commented:
@gavanderhoorn @fmauch @ipa-nhg @miguelprada 

Hey,

I think it's ready ? Can some of you review it when you get some time ? If needed I can reshape the commits to be easier to review (but someone has to guide me, I'm not sure how I can separate them differently).

Thank you,
Bests,
Ludovic

fmauch commented on ur_description/urdf/ur.common.xacro:
```diff
@@ -0,0 +1,160 @@
+    </inertial>
+  </xacro:macro>
+
+  <xacro:macro name="read_file" params="joints_limits_parameters_file physical_parameters_file visual_parameters_file">
```
Maybe I would change the name to something more meaningful like `read_model_data`, as `read_file` sounds like it is just reading one file (which was I think the original idea) but does not give much context.

fmauch commented on ur_description/urdf/ur.geometry.xacro:
```diff
@@ -0,0 +1,254 @@
+          <mesh filename="${upper_arm_collision_mesh}"/>
+        </geometry>
+      </collision>
+      <xacro:cylinder_inertial radius="${upperarm_inertia_radius}" length="${-a2}" mass="${upper_arm_mass}">
```
We could use the calculated lengths here:

```suggestion
      <xacro:cylinder_inertial radius="${upperarm_inertia_radius}" length="${upper_arm_length}" mass="${upper_arm_mass}">
```

fmauch commented on ur_description/urdf/ur.geometry.xacro:
```diff
@@ -0,0 +1,254 @@
+        </geometry>
+      </collision>
+      <xacro:cylinder_inertial radius="${upperarm_inertia_radius}" length="${-a2}" mass="${upper_arm_mass}">
+        <origin xyz="0.0 ${-upper_arm_inertia_offset} ${-a2/2.0}" rpy="0 0 0" />
```
```suggestion
        <origin xyz="0.0 ${-upper_arm_inertia_offset} ${upper_arm_length/2.0}" rpy="0 0 0" />
```

fmauch commented on ur_description/urdf/ur.geometry.xacro:
```diff
@@ -0,0 +1,254 @@
+          <mesh filename="${forearm_collision_mesh}"/>
+        </geometry>
+      </collision>
+      <xacro:cylinder_inertial radius="${forearm_inertia_radius}" length="${-a3}"  mass="${forearm_mass}">
```
```suggestion
      <xacro:cylinder_inertial radius="${forearm_inertia_radius}" length="${forearm_length}"  mass="${forearm_mass}">
```

fmauch commented on ur_description/urdf/ur.geometry.xacro:
```diff
@@ -0,0 +1,254 @@
+        </geometry>
+      </collision>
+      <xacro:cylinder_inertial radius="${forearm_inertia_radius}" length="${-a3}"  mass="${forearm_mass}">
+        <origin xyz="0.0 0.0 ${-a3/2.0}" rpy="0 0 0" />
```
```suggestion
        <origin xyz="0.0 0.0 ${forearm_length/2.0}" rpy="0 0 0" />
```

fmauch commented on ur_bringup/launch/ur5_bringup.launch:
```diff
@@ -7,17 +7,17 @@
-  <arg name="limited" default="false" doc="If true, limits joint range [-PI, PI] on all joints." />
   <arg name="min_payload"  default="0.0" doc="lower payload bound for the robot" />
-  <arg name="max_payload"  default="10.0" doc="upper payload bound for the robot" />
+  <arg name="max_payload"  default="20.0" doc="upper payload bound for the robot" />
```
Is that change deliberate?
```suggestion
  <arg name="max_payload"  default="10.0" doc="upper payload bound for the robot" />
```

fmauch commented on ur_bringup/launch/ur3_bringup.launch:
```diff
@@ -7,17 +7,17 @@
-  <arg name="limited" default="false" doc="If true, limits joint range [-PI, PI] on all joints."/>
   <arg name="min_payload"  default="0.0" doc="lower payload bound for the robot" />
-  <arg name="max_payload"  default="3.0" doc="upper payload bound for the robot" />
+  <arg name="max_payload"  default="20.0" doc="upper payload bound for the robot" />
```
Copy&Paste error?
```suggestion
  <arg name="max_payload"  default="3.0" doc="upper payload bound for the robot" />
```

fmauch commented on ur_description/config/ur10/physical_parameters.yaml:
```diff
@@ -0,0 +1,77 @@
+  wrist_2_mass: 1.96
+  wrist_3_mass: 0.202
+
+  shoulder_radius: x0.060   # FROM UR5 DON'T USE
```
We don't use these anywhere, so we could also drop them, right? This could go for all models, I think.

fmauch reviewed (commented):
The e-series is deliberately left out of this MR? We could also have the e-series do the same thing or even reuse the same ur_common and ur_geometry files.

One minor addition: In #435 we added an rviz config to the `cfg` folder. Now we are saving robot data in a `config` folder. I think, we should merge those folders together.

One more addition: Removing the joint_limited launch files is a great thing, but this will be a breaking change for a lot of people, right?

dhled commented:
Thanks @fmauch for your comments.

> The e-series is deliberately left out of this MR? We could also have the e-series do the same thing or even reuse the same ur_common and ur_geometry files.

I will merge the #441 directly in this one so we will have both.

> One minor addition: In #435 we added an rviz config to the cfg folder. Now we are saving robot data in a config folder. I think, we should merge those folders together.

Ok

> One more addition: Removing the joint_limited launch files is a great thing, but this will be a breaking change for a lot of people, right?

Yes. That's why I suggested to do that for melodic only. I might suggest to go even further by having a single launch file and 6 yaml file pointing to the config yaml files of each of the model (to try avoiding the duplicating copy paste errors :P )  

Regarding you changes, I'm on it. 

For gazebo, I might suggest a solution using a xacro argument in the xacro_robot file that will be set in gazebo launch file adding the missing link to "elevate" the robot.

gavanderhoorn commented:
> For gazebo, I might suggest a solution using a xacro argument in the xacro_robot file that will be set in gazebo launch file adding the missing link to "elevate" the robot.

Let's not do that.

I propose we completely factor out the Gazebo-specific parts of the `.xacro`s and then add the Gazebo-infrastructure and necessary files to `ur_gazebo` (similar to [abb_irb120_gazebo](https://github.com/ros-industrial/abb_experimental/tree/35af6b1a359777052292dfae834535d79bc4d725/abb_irb120_gazebo)).

dhled commented:
Note to me: Integrate #437

Commit:
create parameters yaml files
* one for each models

Commit:
made common macro for ur_robot urdf
* removed each model specific xacro
* use of yaml files
* pass yaml files as parameters
* common ur_robo macro
* remved ur_gazebo specific parts

Commit:
update ur_description launch files
* use of yaml files parameters
* added e_series
* create a common launch file to avoir duplicated

Commit:
changed ur_bringup and related tests
* added e-series

Commit:
change gazebo launch file and tests
* added e-series

Commit:
create urdf files for ur_gazebo

dhled commented:
@gavanderhoorn Can you update the melodic devel with the last change of the kinetic devel ? (I have integrated them)

@fmauch @gavanderhoorn I think I have integrated all your comments:
* change in geometry description to a2/a3 converted in the real lenght name
* change in the bringup payload
* moved gazebo part to ur_gazebo
* created xacro macro and urdf.xacro files to ur_gazebo to make the simulation work

Added:
* deleted all ur models specific files (except configuration yaml and launch files)
* some common_launch files to avoid duplicated code

Not Sure:
* I have kept ur.transmission xacro in the ur_description (I have understood this can be use by ros_control).

I have rebased with comits and description, which I hope will help review the PR.

dhled commented:
Rebased on the current melodic-devel branch

gavanderhoorn commented:
> Rebased on the current melodic-devel branch

thanks. Apologies for breaking the PR :frowning_face:

dhled commented:
I fixed error in the meta package description (there was still ur_e_gazebo / ur_e_description)

fmauch commented:
Is there anything holding this back from merging? With this being merged, we could push forward #414 so people would not have to use my fork when using the ur_robot_driver anymore.

dhled commented:
@ipa-nhg / @ipa-hsd, if you have some time to review this one ?

miguelprada commented on ur_bringup/launch/ur10_bringup.launch:
```diff
@@ -11,15 +10,19 @@
+  <!-- robot model parameters files-->
+  <arg name="joints_limits_params" default="$(find ur_description)/config/ur10/joints_limits_parameters.yaml"/>
+  <arg name="physical_params" default="$(find ur_description)/config/ur10/physical_parameters.yaml"/>
+
```
Any particular reason not to expose the `visual_params` argument to the `ur..._upload.launch` files?

dhled replied:
not really (or at least I don't remember why I did that) ... I will change that

miguelprada commented on ur_description/launch/view_ur10e.launch:
```diff
@@ -1,10 +1,10 @@
   <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" >
     <param name="use_gui" value="true"/>
   </node>
   <node name="robot_state_publisher" pkg="robot_state_publisher" type="state_publisher" />
```
Not quite part of this PR, but this should probably be replaced to use `robot_state_publisher` instead of the deprecated `state_publisher`.

miguelprada commented on ur_description/launch/ur_common.launch:
```diff
@@ -0,0 +1,22 @@
+  <arg name="safety_pos_margin" default="0.15" doc="The lower/upper limits in the safety controller" />
+  <arg name="safety_k_position" default="20" doc="Used to set k position in the safety controller" />
+
+  <param name="robot_description" command="$(find xacro)/xacro '$(find ur_description)/urdf/ur_robot.urdf.xacro' --inorder 
```
Since this targets `melodic-devel`, shouldn't `--inorder` be dropped?

miguelprada reviewed (commented):
In addition to the, admittedly minor, inline comments, I think this is good to go. Particularly considering it's probably been in use for a while already, via @fmauch's fork.

Apologies for letting this sit idle for so long, @ipa-led.

Commit:
added visual parameters to bringup launch

Commit:
removed --inorder for common launch
* default in melodic

Commit:
used robot_state_publisher instead of state_publisher

dhled commented:
@fmauch @miguelprada @gavanderhoorn 

I should have added feature to solve Miguel's comment (sorry, took me way too long).

gavanderhoorn commented:
@ipa-led: picking this up again. I've just changed the target to a new `melodic-devel-staging`, to avoid users using `melodic-devel` and finding it in a broken state (we have a nr of other changes coming in from #448).

We'll merge this as-is now.

Thanks for iterating and for the (enormous) amount of effort. I believe this is definitely an improvement over the current state of things. We trade `N` almost identical macros for a single one with a nr of arguments. I believe that will scale better wrt maintenance, but let's see.

gavanderhoorn commented:
@ipa-led: could you clarify why you've chosen to separate the parameters for the various parts of the models into different files? There are different files for the appearance and physical parameters. I'm trying to understand whether it would make sense to merge those two.

The joint limits can certainly be kept separate: if we restructure/extend the file a little bit, it could also be used directly by MoveIt (instead of the separate file that's now hosted in the `config/` sub dir of the MoveIt cfg packages).

dhled commented:
> 
> 
> @ipa-led: could you clarify why you've chosen to separate the parameters for the various parts of the models into different files? There are different files for the appearance and physical parameters. I'm trying to understand whether it would make sense to merge those two.

I think at that time I have separated by "concerns" (graphical, physicial, limits) but maybe it doesn't make sense to split. It just that maybe people wants to modify one file and ensuring that it doesn't impact the others.

> The joint limits can certainly be kept separate: if we restructure/extend the file a little bit, it could also be used directly by MoveIt (instead of the separate file that's now hosted in the `config/` sub dir of the MoveIt cfg packages).

I will have a look at that :)

gavanderhoorn commented:
You don't have to add anything to this PR. We'll merge into the staging branch and can then continue there with follow-up PRs.

fmauch commented:
Anything happening here recently?

lianghongzhuo commented:
Can we please push this forward?

dhled commented:
> 
> 
> Anything happening here recently?

I don't know if we are waiting for this be merge or for me to add other things.

v4hn commented:
@gavanderhoorn *twice* now you wrote you will merge this as-is, but that was two months ago.
Are you going to do so?

miguelprada commented:
Sorry, I had assumed this was already merged. Apologies for letting it sit for so long.

Any reason not to merge already, @gavanderhoorn, @ipa-nhg?

lianghongzhuo commented:
What's going on here?

ipa-nhg reviewed (approved):
LGTM :+1: Thanks @ipa-led

gavanderhoorn reviewed (approved):
Merging this into staging branch.

gavanderhoorn commented:
Thanks @ipa-led :+1: 

Great work and I believe this will greatly simplify maintenance of these packages in the future.

:robot: :fr:
